### PR TITLE
Add unsafe_string for Cwstring

### LINKED ISSUE #28863: No `unsafe_string` for `Cwstring`
There is no obvious way to convert `Cwstring` to a Julia `String`:
```
julia> s = "aaa";

julia> a = Base.cconvert(Cwstring,s);

julia> w = Base.unsafe_convert(Cwstring, a)
Cwstring(0x000000010e4d5fd0)

julia> unsafe_string(w)
ERROR: MethodError: no method matching unsafe_string(::Cwstring)
Closest candidates are:
  unsafe_string(::Cstring) at c.jl:193
  unsafe_string(::Union{Ptr{Int8}, Ptr{UInt8}}) at strings/string.jl:56
  unsafe_string(::Union{Ptr{Int8}, Ptr{UInt8}}, ::Integer) at strings/string.jl:52
```

from https://stackoverflow.com/questions/51994365/convert-refcwstring-to-string

### PR DESCRIPTION
**Pull request by simonbyrne**
Also adds some tests for Cwstring handling, fixes #28863.

### CODE REVIEW

**Comment by simonbyrne** [comment]
Not sure why the `ccall` in the test is failing on Windows. When I try it in a VM, I get the following error:
```
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ffc90856807 -- wcstombs at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
in expression starting at REPL[4]:1
wcstombs at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
swprintf at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
top-level scope at .\REPL[4]:1
jl_toplevel_eval_flex at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:834
jl_toplevel_eval_flex at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:790
jl_toplevel_eval at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:849 [inlined]
jl_toplevel_eval_in at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:883
eval at .\boot.jl:331
eval_user_input at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:134
repl_backend_loop at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:195
start_repl_backend at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:180
#run_repl#37 at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:292
run_repl at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:288
#802 at .\client.jl:398
jfptr_YY.802_44758.clone_1 at C:\Users\IEUser\AppData\Local\Programs\Julia 1.6.0-DEV\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\julia.h:1700 [inlined]
do_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:655
jl_f__apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:669 [inlined]
jl_f__apply_latest at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:705
#invokelatest#1 at .\essentials.jl:710 [inlined]
invokelatest at .\essentials.jl:709 [inlined]
run_main_repl at .\client.jl:382
exec_options at .\client.jl:312
_start at .\client.jl:487
jfptr__start_45990.clone_1 at C:\Users\IEUser\AppData\Local\Programs\Julia 1.6.0-DEV\lib\julia\sys.dll (unknown line)
unknown function (ip: 00000000004017E1)
unknown function (ip: 0000000000401BD6)
unknown function (ip: 00000000004013DE)
unknown function (ip: 000000000040151A)
BaseThreadInitThunk at C:\Windows\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
Allocations: 839846 (Pool: 839578; Big: 268); GC: 1
```

@vtjnash any ideas?

File: base/c.jl [hunk 1]
```diff
@@ -414,6 +414,18 @@ function transcode(::Type{UInt8}, src::AbstractVector{UInt16})
     return dst
 end
 
+function unsafe_string(p::Ptr{T}, length::Integer) where {T<:Union{UInt16,UInt32,Cwchar_t}}
+    transcode(String, unsafe_wrap(Array, p, length; own=false))
+end
+function unsafe_string(p::Ptr{T}) where {T<:Union{UInt16,UInt32,Cwchar_t}}
+    n = 1
+    while unsafe_load(p, n) != 0
+        n+=1
```
Comment (vtjnash) [inline base/c.jl:423]:
```suggestion
        n += 1
```

File: base/c.jl [hunk 2]
```diff
@@ -414,6 +414,18 @@ function transcode(::Type{UInt8}, src::AbstractVector{UInt16})
...
+    end
+    unsafe_string(p,n-1)
```
Comment (vtjnash) [inline base/c.jl:425]:
```suggestion
    return unsafe_string(p, n - 1)
```

**Review by vtjnash — COMMENTED** [review]
This symbol on Windows is renamed `_swprintf`

**Comment by simonbyrne** [comment]
I see the same thing with `_swprintf`:
```
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.0-DEV.109 (2020-05-27)
 _/ |\__'_|_|_|\__'_|  |  Commit 806dfca07d (0 days old master)
|__/                   |

julia> len = @ccall _swprintf(Array{Cwchar_t}(undef,100)::Ptr{Cwchar_t}, 100::Csize_t, "aa"::Cwstring)::Cint

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ffa64e16807 -- wcstombs at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
in expression starting at REPL[1]:1
wcstombs at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
swprintf at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
top-level scope at .\REPL[1]:1
jl_toplevel_eval_flex at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:834
jl_toplevel_eval_flex at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:790
jl_toplevel_eval at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:849 [inlined]
jl_toplevel_eval_in at /cygdrive/d/buildbot/worker/package_win64/build/src\toplevel.c:883
eval at .\boot.jl:331
eval_user_input at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:134
repl_backend_loop at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:195
start_repl_backend at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:180
#run_repl#37 at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:292
run_repl at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\REPL\src\REPL.jl:288
#802 at .\client.jl:398
jfptr_YY.802_44758.clone_1 at C:\Users\IEUser\AppData\Local\Programs\Julia 1.6.0-DEV\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\julia.h:1700 [inlined]
do_apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:655
jl_f__apply at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:669 [inlined]
jl_f__apply_latest at /cygdrive/d/buildbot/worker/package_win64/build/src\builtins.c:705
#invokelatest#1 at .\essentials.jl:710 [inlined]
invokelatest at .\essentials.jl:709 [inlined]
run_main_repl at .\client.jl:382
exec_options at .\client.jl:312
_start at .\client.jl:487
jfptr__start_45990.clone_1 at C:\Users\IEUser\AppData\Local\Programs\Julia 1.6.0-DEV\lib\julia\sys.dll (unknown line)
unknown function (ip: 00000000004017E1)
unknown function (ip: 0000000000401BD6)
unknown function (ip: 00000000004013DE)
unknown function (ip: 000000000040151A)
BaseThreadInitThunk at C:\Windows\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
Allocations: 2541 (Pool: 2532; Big: 9); GC: 0
```

**Comment by simonbyrne** [comment]
Interestingly, `swprintf_s` seems to work correctly, except for float printing.

**Comment by simonbyrne** [comment]
Actually, float printing is broken for `sprintf` as well:
```
julia> buffer = Array{UInt8}(undef, 100);

julia> len = @ccall sprintf(buffer::Ptr{Cchar}, "aa %f"::Cstring; 1.2f0::Cfloat)::Cint
4

julia> transcode(String, buffer[1:len])
"aa f"
```
@vtjnash Is there something up with the calling convention?

File: base/c.jl [hunk 3]
```diff
@@ -414,6 +414,18 @@ function transcode(::Type{UInt8}, src::AbstractVector{UInt16})
     return dst
 end
 
+function unsafe_string(p::Ptr{T}, length::Integer) where {T<:Union{UInt16,UInt32,Cwchar_t}}
+    transcode(String, unsafe_wrap(Array, p, length; own=false))
+end
+function unsafe_string(p::Ptr{T}) where {T<:Union{UInt16,UInt32,Cwchar_t}}
```
[the remark this answers is not in the document]
Comment (simonbyrne) [reply base/c.jl]: Well, I've removed it. Can always add it back if there is demand.

**Comment by vtjnash** [comment]
Ah, I see, if you read the documentation closely, it specifies that the `count` parameter is part of a future API for `swprintf` and may or may not be assumed to actually be present currently (in particular, for msvcrt, it is not present). If you remove `count`, all works.

> swprintf conforms to the ISO C Standard, which requires the second parameter, count, of type size_t. To force the old nonstandard behavior, define _CRT_NON_CONFORMING_SWPRINTFS. In a future version, the old behavior may be removed, so code should be changed to use the new conformant behavior.

**Pull request merged**